component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java

```diff
@@ -395,6 +396,33 @@ private void checkTotpEnabled(AuthenticationContext context, String username) th
         }
     }
 
+    /**
+     * Update the secret key in the context with value in database before validate.
+     *
+     * @param context  Authenticated context.
+     * @param username Authenticated users' username
+     * @throws AuthenticationFailedException When getting the secret key from database.
+     */
+    private void checkForUpdatedSecretKey(AuthenticationContext context, String username)
+            throws AuthenticationFailedException {
+
+        try {
+            String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(username);
+            UserRealm userRealm = TOTPUtil.getUserRealm(username);
+            if (userRealm != null) {
+                Map<String, String> userClaimValues = userRealm.getUserStoreManager()
+                        .getUserClaimValues(tenantAwareUsername,
+                                new String[]{TOTPAuthenticatorConstants.SECRET_KEY_CLAIM_URL}, null);
+                String secretKey = userClaimValues.get(TOTPAuthenticatorConstants.SECRET_KEY_CLAIM_URL);
+                if (StringUtils.isNotEmpty(secretKey)) {
```

Comment:
Do we need to check for a blank here?

Reply:
Yes. This flow is only for the users that have no secret key available in database when starting the authentication flow. Therefore we keep the generated secret key in the context, only if database get change from some other process we override the context info.